src/test/java/io/confluent/examples/streams/CustomStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,538 @@
+      new KeyValueWithTimestamp<>("alice", 555.55, TimeUnit.MILLISECONDS.toMillis(30)),
+      new KeyValueWithTimestamp<>("alice", 666.66, TimeUnit.MILLISECONDS.toMillis(40))
+    );
+
+    final List<KeyValueWithTimestamp<String, Long>> inputTableRecords = Arrays.asList(
+      new KeyValueWithTimestamp<>("alice", 1L, TimeUnit.MILLISECONDS.toMillis(20)),
+      new KeyValueWithTimestamp<>("alice", 2L, TimeUnit.MILLISECONDS.toMillis(39)),
+      new KeyValueWithTimestamp<>("bobby", 8L,
+        approxMaxWaitTimePerRecordForTableData.plus(Duration.ofSeconds(1)).toMillis())
+    );
+
+    final List<KeyValue<String, Pair<Double, Long>>> expectedOutputRecords = Arrays.asList(
+      new KeyValue<>("alice", new Pair<>(999.99, 1L)),
+      new KeyValue<>("alice", new Pair<>(555.55, 1L)),
+      new KeyValue<>("alice", new Pair<>(666.66, 2L)),
+      new KeyValue<>("bobby", new Pair<>(222.22, null))
+    );
+
+    //
+    // Step 1: Define and start the processor topology.
+    //
+    final StreamsBuilder builder = new StreamsBuilder();
+
+    final Properties streamsConfiguration = new Properties();
+    streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "custom-join-integration-test");
+    streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+    streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+    // Use a temporary directory for storing state, which will be automatically removed after the test.
+    streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
+
+    // This state store is used to temporarily buffer any records arriving at the stream side of the join, so that
+    // we can wait (if needed) for matching data to arrive at the table side.
+    final StoreBuilder<KeyValueStore<String, Pair<Double, Instant>>> streamBufferStateStore =
+      Stores
+        .keyValueStoreBuilder(
+          Stores.persistentKeyValueStore("stream-buffer-state-store"),
+          Serdes.String(),
+          new PairSerde<>(Serdes.Double(), new InstantSerde())
+        )
+        .withCachingEnabled();
+    builder.addStateStore(streamBufferStateStore);
+
+    // Read the input data.
+    final KStream<String, Double> stream = builder.stream(inputTopicForStream, Consumed.with(Serdes.String(), Serdes.Double()));
+    final KTable<String, Long> table =
+      builder.table(inputTopicForTable, Consumed.with(Serdes.String(), Serdes.Long()),
+        // Disabling this state store's cache ensures that table-side data is seen more immediately and that reading the
+        // stream-side data doesn't advance the stream-time too quickly (which might result in join output being
+        // sent by the stream side before the processing topology had the chance to look at the table side).
+        Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as(tableStoreName).withCachingDisabled());
```

Comment:
The `withCachingDisabled()` is required for this integration test to pass.  Attempting to achieve the same effect by setting `max.task.idle.ms` to non-zero values did NOT work. cc @mjsax.